File: 700/764 - Largest Plus Sign.py
class Solution:
    def orderOfLargestPlusSign(self, N, mines):
        """
        :type N: int
        :type mines: List[List[int]]
        :rtype: int
        """
        mines_set = set([(m[0], m[1],) for m in mines])
        dp = [[0 for _ in range(N)] for _ in range(N)]
        ret = 0
        
        for i in range(N):
            count = 0
            for j in range(N):
                count = 0 if (i, j) in mines_set else count + 1
                dp[i][j] = count
            
            count = 0
            for j in range(N - 1, -1, -1):
                count = 0 if (i, j) in mines_set else count + 1
                dp[i][j] = min(count, dp[i][j])
            
        for i in range(N):
            count = 0
            for j in range(N):
                count = 0 if (j, i) in mines_set else count + 1
                dp[j][i] = min(count, dp[j][i])
                
            count = 0
            for j in range(N - 1, -1, -1):
                count = 0 if (j, i) in mines_set else count + 1
                dp[j][i] = min(count, dp[j][i])
                ret = max(ret, dp[j][i])
        return ret

File: csharp/764.largest-plus-sign.cs
/*
 * @lc app=leetcode id=764 lang=csharp
 *
 * [764] Largest Plus Sign
 */

using System;
using System.Collections.Generic;

// @lc code=start
public class Solution {
    public int OrderOfLargestPlusSign(int N, int[][] mines) {
        int[,] dp = new int[N, N];
        int ret = 0;
        var set = new HashSet<int>(mines.Select(loc => loc[0] * N + loc[1]));
        for (int i = 0; i < N; ++i) {
            int count = 0;
            for (int j = 0; j < N; ++j) {
                count = set.Contains(i * N + j) ? 0 : (1 + count);
                dp[i, j] = count;
            }
            count = 0;
            for (int j = N - 1; j >= 0; --j) {
                count = set.Contains(i * N + j) ? 0 : (1 + count);
                dp[i, j] = Math.Min(count, dp[i, j]);
            }
        }
        for (int j = 0; j < N; ++j) {
            int count = 0;
            for (int i = 0; i < N; ++i) {
                count = set.Contains(i * N + j) ? 0 : (1 + count);
                dp[i, j] = Math.Min(count, dp[i, j]);
            }
            count = 0;
            for (int i = N - 1; i >= 0; --i) {
                count = set.Contains(i * N + j) ? 0 : (1 + count);
                dp[i, j] = Math.Min(count, dp[i, j]);
                ret = Math.Max(ret, dp[i, j]);
            }
        }
        return ret;
    }
}
// @lc code=end

